src/printer/embed.ts

```diff
@@ -19,135 +19,143 @@ const {
 type supportedStyleLang = 'css' | 'scss' | 'sass';
 
 // https://prettier.io/docs/en/plugins.html#optional-embed
-export function embed(
-	path: AstPath,
-	print: printFn,
-	textToDoc: (text: string, options: object) => Doc,
-	opts: ParserOptions
-) {
-	const node = path.getValue();
+type Embed =
+	| ((
+			// Parses and prints the passed text using a different parser.
+			// You should set `options.parser` to specify which parser to use.
+			textToDoc: TextToDoc,
+			// Prints the current node or its descendant node with the current printer
+			print: printFn,
+			// The following two arguments are passed for convenience.
+			// They're the same `path` and `options` that are passed to `embed`.
+			path: AstPath,
+			options: ParserOptions
+	  ) => Promise<Doc | undefined> | Doc | undefined)
+	| Doc
+	| undefined;
+
+type TextToDoc = (text: string, options: ParserOptions) => Promise<Doc>;
 
-	if (!node) return null;
+// https://prettier.io/docs/en/plugins.html#optional-embed
+export function embed(path: AstPath, opts: ParserOptions): Embed {
+	return async (textToDoc, print) => {
```

Comment:
Suggest move some of these node type check  out of the async function, so Prettier don't need await for all nodes, the print process will be faster.